pkg/mutation/schema/schema_test.go

```diff
@@ -0,0 +1,241 @@
+)
+
+var _ MutatorWithSchema = &mockMutator{}
+
+type mockMutator struct {
+	id        types.ID
+	ForceDiff bool
+	Bindings  []Binding
+	path      string
+	pathCache *parser.Path
+}
+
+func (m *mockMutator) Matches(obj metav1.Object, ns *corev1.Namespace) bool { return false }
+
+func (m *mockMutator) Mutate(obj *unstructured.Unstructured) error { return nil }
+
+func (m *mockMutator) ID() types.ID { return m.id }
+
+func (m *mockMutator) HasDiff(other types.Mutator) bool {
+	if m.ForceDiff {
+		return true
+	}
+	return !reflect.DeepEqual(m, other)
+}
+
+func deepCopyBindings(bindings []Binding) []Binding {
+	cpy := []Binding{}
+	for _, b := range bindings {
+		cpy = append(cpy, Binding{
+			Groups:   append([]string{}, b.Groups...),
+			Kinds:    append([]string{}, b.Kinds...),
+			Versions: append([]string{}, b.Versions...),
+		})
+	}
+	return cpy
+}
+
+func (m *mockMutator) DeepCopy() types.Mutator {
+	return &mockMutator{
+		id:        m.id,
+		ForceDiff: m.ForceDiff,
+		Bindings:  deepCopyBindings(m.Bindings),
+		path:      m.path,
+	}
+}
+
+func (m *mockMutator) SchemaBindings() []Binding { return m.Bindings }
+
+func (m *mockMutator) Path() *parser.Path {
+	if m.pathCache != nil {
+		return m.pathCache
+	}
+	out, err := parser.Parse(m.path)
+	if err != nil {
+		panic(err)
+	}
+	m.pathCache = out
+	return out
+}
+
+func id(name string) types.ID {
+	return types.ID{Name: name}
+}
+
+func bindings(kind string) []Binding {
+	return []Binding{{Groups: []string{""}, Versions: []string{"v1"}, Kinds: []string{kind}}}
+}
+
+func gvk(kind string) schema.GroupVersionKind {
+	return schema.GroupVersionKind{
+		Group:   "",
+		Version: "v1",
+		Kind:    kind,
+	}
+}
+
+func sp(s string) *string {
+	return &s
+}
+
+func simpleMutator(mid, kind, path string) *mockMutator {
+	return &mockMutator{
+		id:       id(mid),
+		Bindings: bindings(kind),
+		path:     path,
+	}
+}
+
+const (
+	upsert = "upsert"
+	remove = "remove"
+)
+
+type testOp struct {
+	op            string
+	errorExpected bool
+	id            types.ID
+	mutator       *mockMutator
+}
+
+// TODO write more tests
+func TestParser(t *testing.T) {
+	tests := []struct {
+		name             string
+		ops              []testOp
+		expectedMutators map[types.ID]MutatorWithSchema
+		expectedSchemas  map[schema.GroupVersionKind]*scheme
+	}{
+		{
+			name: "Simple upsert",
+			ops: []testOp{
+				{
+					op:      upsert,
+					mutator: simpleMutator("simple", "FooKind", "spec.someValue"),
+				},
+			},
+			expectedMutators: map[types.ID]MutatorWithSchema{
+				id("simple"): simpleMutator("simple", "FooKind", "spec.someValue"),
+			},
+			expectedSchemas: map[schema.GroupVersionKind]*scheme{
+				gvk("FooKind"): {
+					gvk: gvk("FooKind"),
+					root: &node{
+						referenceCount: 1,
+						nodeType:       parser.ObjectNode,
+						children: map[string]*node{
+							"spec": {
+								referenceCount: 1,
+								nodeType:       parser.ObjectNode,
+								children: map[string]*node{
```

Comment:
This reveals a bug:

If we allow leaf nodes with undeclared types from one path, a deeper path may define that type.

If the deeper path subsequently gets deleted, that type inference will not be removed, even though the implicit schemas for currently recognized mutation don't support the type inference.

This could lead us to reject otherwise valid mutation additions.

Reply:
I think the leaf always have a type (which happens to be "object").
What could happen is that a deeper path defines the node as a list, then we remove it and if the node becomes a leaf again, it should change type.

Actually, this is also true when adding. If we have spec.containers as a leaf, it will be of type object.
If we then try to add spec.containers[name: foo].something, the type check in https://github.com/open-policy-agent/gatekeeper/pull/979/files#diff-4dc841aafbd760fba8fe37a39b84bea27b99d3065a948a381fcb4bca1e5699adR256 will fail.

Would it make sense to add a new type "leaf" here, compatible with both object and list, with the type changing back and forth if the node is promoted from "leaf" to field inside the path or viceversa?

Reply:
That would work, though I think it might be slightly less complex and use less space if we simply ignore the last Path() entry for each mutation object. It should have the same effect, but there will be fewer objects to keep track of.

Reply:
Ah, neat. In that case we just ignore the type if it is a leaf, and we trust the type if it's in the middle of a path.

Reply:
yep!

I implemented the fix. It's a bit weird because having a list implies the next entry __must__ be an object.